### PR TITLE
feat(autoware_trajectory_optimizer): mpt plugin 

### PR DESCRIPTION
**Pull request by danielsanchezaran**
# PR Description - MPT Optimizer Plugin

## Description

This PR adds a new MPT (Model Predictive Trajectory) optimizer plugin to the trajectory optimizer package. The plugin refines trajectory geometry using model predictive optimization with bicycle kinematics, leveraging the MPT solver from `autoware_path_optimizer`.

**What it does:**

- Optimizes trajectory geometry within dynamically-generated corridor bounds
- Uses bicycle kinematics model with steering constraints for feasibility
- Recalculates trajectory dynamics (acceleration, time stamps) after geometric changes
- Provides map-independent operation by generating bounds from trajectory geometry

**Why it's needed:**

- Designed for learning-based planner outputs (e.g., diffusion planner) that may need geometric refinement
- Model predictive approach considers vehicle kinematics over prediction horizon (vs local smoothing)
- Reduces dependency on lanelet map by generating geometric bounds from trajectory
- Complements existing smoothers with kinematically-aware optimization

**Status:**

- Disabled by default (`use_mpt_optimizer: false`) - experimental feature
- Known limitation: Struggles with sharp in-cutting curves
- Requires further validation before production use

## Related links

**Parent Issue:**

- <https://tier4.atlassian.net/browse/T4DEV-39886>

**Private Links:**

- [TIER IV Demo Videos and Technical Details](https://tier4.atlassian.net/wiki/spaces/ATT/pages/4565303548/MPT+plugin+for+the+Optimizer)

## How was this PR tested?

**Unit Tests:**

- Added 21 comprehensive unit tests for MPT optimizer utilities
- All tests passing (39/39 total package tests)
- Tests cover:
  - Acceleration calculation using kinematic equations
  - Time interval calculation with edge cases
  - Curvature estimation
  - Corridor width adaptation
  - Bound generation
  - Full dynamics recalculation pipeline

**Simulation Tests:**

- Tested in urban area maps (see private links)
- Tested on test course with large curves
- Validated with diffusion planner trajectories

### Plugin Ordering

The MPT optimizer is placed **after all geometric smoothing** but **before velocity optimization**:

```yaml
plugin_names:
  - TrajectoryPointFixer
  - TrajectoryKinematicFeasibilityEnforcer
  - TrajectoryQPSmoother
  - TrajectoryKinematicFeasibilityEnforcer  
  - TrajectoryEBSmootherOptimizer
  - TrajectorySplineSmoother
  - TrajectoryMPTOptimizer                  
  - TrajectoryVelocityOptimizer            
  - TrajectoryExtender
```

**Rationale:**

- MPT requires trajectories with low yaw noise (previous smoothers provide this)
- MPT destroys acceleration information during optimization
- Velocity optimizer runs after to properly optimize velocity profile and limits

### Disabled by Default

The plugin is experimental and disabled by default (`use_mpt_optimizer: false`) because:

- Performance needs further validation
- Struggles with certain sharp curves where ego performs in-cutting maneuvers
- Computational cost (~3ms) may not be suitable for all applications

### Code Organization

New directory structure for plugin utilities:

```
trajectory_optimizer_plugins/
├── plugin_utils/  (NEW)
│   ├── trajectory_mpt_optimizer_utils.hpp/cpp
│   └── trajectory_point_fixer_utils.hpp/cpp (moved)
```

#### Additions

| Parameter Name | Type | Default Value | Description |
|:---------------|:-----|:--------------|:------------|
| `use_mpt_optimizer` | `bool` | `false` | Enable MPT optimizer plugin (disabled by default - experimental) |
| `trajectory_mpt_optimizer.corridor_width_m` | `double` | `3.5` | Base perpendicular offset distance from trajectory [m] |
| `trajectory_mpt_optimizer.enable_adaptive_width` | `bool` | `true` | Enable width adjustments based on curvature and velocity |
| `trajectory_mpt_optimizer.curvature_width_factor` | `double` | `0.5` | Additional width per unit curvature [m/rad] |
| `trajectory_mpt_optimizer.velocity_width_factor` | `double` | `0.3` | Additional width scaling at low speeds [m] |
| `trajectory_mpt_optimizer.min_clearance_m` | `double` | `0.5` | Minimum clearance from vehicle edges [m] |
| `trajectory_mpt_optimizer.reset_previous_data_each_iteration` | `bool` | `true` | Reset MPT warm start data each cycle |
| `trajectory_mpt_optimizer.enable_debug_info` | `bool` | `false` | Publish debug visualization markers |
| `trajectory_mpt_optimizer.output_delta_arc_length_m` | `double` | `1.0` | Point spacing in optimized trajectory [m] |
| `trajectory_mpt_optimizer.output_backward_traj_length_m` | `double` | `5.0` | Trajectory extension backward from ego [m] |
| `trajectory_mpt_optimizer.ego_nearest_dist_threshold_m` | `double` | `3.0` | Distance threshold for ego-trajectory matching [m] |
| `trajectory_mpt_optimizer.ego_nearest_yaw_threshold_deg` | `double` | `45.0` | Yaw threshold for ego-trajectory matching [deg] |
| `trajectory_mpt_optimizer.acceleration_moving_average_window` | `int` | `5` | Moving average window size for acceleration smoothing |

Plus extensive MPT algorithm parameters (weights, constraints, validation) - see `config/plugins/trajectory_mpt_optimizer.param.yaml`.

## Effects on system behavior

**When disabled (default):** No behavior change.

**When enabled (`use_mpt_optimizer: true`):**

- Additional geometric refinement of trajectories within corridor bounds
- Trajectory positions may differ from previous smoothers
- Accelerations and time stamps recalculated for kinematic consistency
- Added computational cost: ~2ms per trajectory
- May struggle with sharp in-cutting maneuvers (known limitation)

---

### Collision Avoidance Capability

**Note:** The underlying MPT solver supports collision avoidance optimization through drivable area bounds (obstacles removed from drivable area create bounds that avoid collisions). However, this PR does **not implement collision avoidance** because:

- We generate simple perpendicular offsets from trajectory (not actual drivable area)
- No obstacle/predicted object data is provided to the MPT optimizer
- Collision avoidance could be added in future work by integrating obstacle data

The collision-related parameters (`soft_collision_free_weight`, `collision_free_constraints`, etc.) are part of the underlying MPT solver configuration but not actively used for obstacle avoidance in this implementation.

### Code Organization Changes

**New utility structure:**

```
trajectory_optimizer_plugins/
└── plugin_utils/  (NEW)
    ├── trajectory_mpt_optimizer_utils.hpp/cpp
    └── trajectory_point_fixer_utils.hpp/cpp (moved from parent directory)
```

Plugin-specific utilities are now organized in a dedicated `plugin_utils/` subdirectory for better code organization and maintainability.

### CODE REVIEW

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/**:
+  ros__parameters:
+    # MPT (Model Predictive Trajectory) Optimizer Parameters
+    # Optimizes trajectory geometry using QP-based optimization with adaptive corridor bounds
```
Comment (go-sakayori) [inline planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml]:
```suggestion
```

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      velocity_width_factor: 0.3
+      min_clearance_m: 0.5
+      reset_previous_data_each_iteration: true
+      enable_debug_info: false
+      output_delta_arc_length_m: 1.0
+      output_backward_traj_length_m: 5.0
+      ego_nearest_dist_threshold_m: 3.0
+      ego_nearest_yaw_threshold_deg: 45.0
+
+      # Acceleration Smoothing
```
Comment (go-sakayori) [inline planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml]:
```suggestion
```
Since there is another comment below

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      option:
+        steer_limit_constraint: false
+        visualize_sampling_num: 1
+        enable_manual_warm_start: false
+        enable_warm_start: false
+        enable_optimization_validation: false
+
+      common:
+        num_points: 100
+        delta_arc_length: 1.0
```
Comment (go-sakayori) [inline planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml:33]: nits: no unit for this parameter name?

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        yaw_error_rate_weight: 0.0
+        steer_input_weight: 1.0
+        steer_rate_weight: 1.0
+        terminal_lat_error_weight: 100.0
+        terminal_yaw_error_weight: 100.0
+        goal_lat_error_weight: 1000.0
+        goal_yaw_error_weight: 1000.0
+
+      avoidance:
+        max_bound_fixing_time: 1.0
```
Comment (go-sakayori) [inline planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml:52]: nits: no unit for this parameter name? Same sfor some parameters below

File: planning/autoware_trajectory_optimizer/src/trajectory_optimizer_plugins/plugin_utils/trajectory_mpt_optimizer_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  traj_points.front().time_from_start.sec = 0;
+  traj_points.front().time_from_start.nanosec = 0;
+
+  // Calculate acceleration and time for each segment
+  for (size_t i = 1; i < traj_points.size(); ++i) {
+    auto & next_point = traj_points[i];
+    auto & curr_point = traj_points[i - 1];
+
+    // Calculate acceleration from velocity change and distance
+    const auto velocity = static_cast<double>(curr_point.longitudinal_velocity_mps);
```
Comment (go-sakayori) [inline planning/autoware_trajectory_optimizer/src/trajectory_optimizer_plugins/plugin_utils/trajectory_mpt_optimizer_utils.cpp:89]: Does the mpt plugin consider negative value for longitudinal velocity? If not better to write a comment in the `mpt_optimizer.md` as it is written in `kinematic_feasibility_enforcer.md`

File: planning/autoware_trajectory_optimizer/src/trajectory_optimizer_plugins/plugin_utils/trajectory_mpt_optimizer_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    const double new_time = curr_time + dt;
+    next_point.time_from_start = rclcpp::Duration::from_seconds(new_time);
+
+    // Update acceleration
+    curr_point.acceleration_mps2 = static_cast<float>(acceleration);
+  }
+
+  // Apply moving average filter to acceleration
+  if (smoothing_window > 1) {
+    const int window_size = std::max(1, smoothing_window);
```
Comment (go-sakayori) [inline planning/autoware_trajectory_optimizer/src/trajectory_optimizer_plugins/plugin_utils/trajectory_mpt_optimizer_utils.cpp]: I think it's a bit redundant here. Use either of them

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml (hunk 3 above)
Comment (danielsanchezaran) [reply planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml:33]: This is a parameter of the MPT class which comes from the trajectory optimizer package, to change this parameter would  mean changing that package also and other packages that use it.

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml (hunk 4 above)
Comment (danielsanchezaran) [reply planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml:52]: This is a parameter from the MPT class itself, the MPT class handles its own declaration of parameters so changing this would imply having to change the autoware_path_optimizer package too

Comment (go-sakayori) [reply planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml:52]: I got it. Let's leave as it is.

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml (hunk 3 above)
Comment (go-sakayori) [reply planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml:33]: I got it. Let's leave as it is.

File: planning/autoware_trajectory_optimizer/src/trajectory_optimizer_plugins/plugin_utils/trajectory_mpt_optimizer_utils.cpp (hunk 6 above)
Comment (danielsanchezaran) [reply planning/autoware_trajectory_optimizer/src/trajectory_optimizer_plugins/plugin_utils/trajectory_mpt_optimizer_utils.cpp]: Fixed in 8964650

File: planning/autoware_trajectory_optimizer/src/trajectory_optimizer_plugins/plugin_utils/trajectory_mpt_optimizer_utils.cpp (hunk 5 above)
Comment (danielsanchezaran) [reply planning/autoware_trajectory_optimizer/src/trajectory_optimizer_plugins/plugin_utils/trajectory_mpt_optimizer_utils.cpp:89]: I added a check for negative velocity in 8964650, I also added a comment in the md file to explain how this plugin is for moving forward basically in fc3c790

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml (hunk 3 above)
Comment (danielsanchezaran) [reply planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml:33]: These parameters are from the MPT class in the autoware_path_optimizer package, which declares its own hard-coded parameters in its constructor, to change the param names I would need to modify that package which is used in other places too.

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml (hunk 4 above)
Comment (danielsanchezaran) [reply planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml:52]: Same as above

File: planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml (hunk 2 above)
Comment (danielsanchezaran) [reply planning/autoware_trajectory_optimizer/config/plugins/trajectory_mpt_optimizer.param.yaml]: Done
